text/0000-float-semantics.md

```diff
@@ -0,0 +1,305 @@
+The second however [actually fails](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0b4b952929c9ebcd2bd50aee54e6cdf4) on current nightlies in debug mode.
+
+This resolves the last open question blocking floating-point operations in `const fn`.
+When the RFC is accepted, the `const_fn_floating_point_arithmetic` feature gate and the `const fn to_bits` methods can be stabilized.
+
+### Assumptions about floating-point environment
+
+This RFC is primarily concerned with the guarantee Rust provides to its users.
+How exactly those guarantees are achieved is an implementation detail, and not observable when writing pure Rust code.
+However, when mixing Rust with inline assembly, those details *do* become observable.
+To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point control bits must be in exactly the same state as when the inline assembly block was entered.
+This is just an instance of the general principle that it is UB for inline assembly to violate any of the invariants that the Rust compiler relies on when implementing Rust semantics on the target hardware.
+Furthermore, observing the floating-point exception state yields entirely unspecified results: Rust floating-point operations may or may not be executed at the place in the code where they were originally written, and the floating-point status bits can change even if no floating-point operation exists in the source code.
+
+(This is very similar to C without `#pragma STD FENV_ACCESS`.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This RFC is too restrictive for some targets and too vague for some users: it is too vague since NaN signs are still left completely non-deterministic, which is not actually the case on hardware.
+  It is also too restrictive as shown by the following subsection.
+- It [looks](https://reviews.freebsd.org/D33599) like some targets, such FreeBSD, leave the floating-point environment in signal handlers in an unspecified state.
+  This is something that [Linux specifically avoided](https://yarchive.net/comp/linux/fp_state_save.html) since it breaks e.g. the glibc `memcpy` implementation (which uses SSE registers if available).
+  Rust (just like C without `#pragma STD FENV_ACCESS`) cannot currently be used to write signal handlers on such targets.
+  There is little we can do here with the current state of LLVM; and even once LLVM provides the necessary features, these signal handlers will need annotations in the code that tell the compiler about the non-default floating point state.
+  Those targets chose to use a semantics that is hard to support well in a highly optimized language, and there's not much we can do to paper over such target quirks.
+  We can only hope that eventually those targets will decide to provide a reliable floating-point environment.
```

Comment:
I should add that this can be mostly worked around with inline assembly that sets the floating-point environment to the default state Rust accepts. While strictly speaking that would have to happen before any Rust function gets called, practically speaking if this is the first thing the signal handler does, it is very unlikely that the compiler will put a floating-point operation between the function start and the inline assembly block.